### PR TITLE
External test modules support

### PR DESCRIPTION
**Pull request by PavelPunegov**
* Adds support of running multi module tests from `compiler/codegen/`. 
* Small cleanup: removes obsolete `OldKonanTest`
* Use more lazy task creation and configuration

### CODE REVIEW

File: build-tools/src/main/groovy/org/jetbrains/kotlin/KonanTest.groovy [hunk 1]
```diff
@@ -607,29 +473,53 @@ fun runTest() {
...
-            compileList.add(project.file("helpers.kt").absolutePath)
+            compileList*.writeTextToFile()
             try {
-                def exePath = buildExePath()
                 if (enableTwoStageCompilation) {
                     // Two-stage compilation.
-                    def klibPath = "${exePath}.klib"
-                    runCompiler(compileList, klibPath, flags + ["-p", "library"])
-                    runCompiler([], exePath, flags + ["-Xinclude=$klibPath"])
+                    def klibPath = "${executablePath()}.klib"
```
Comment (ilmat192) [inline build-tools/src/main/groovy/org/jetbrains/kotlin/KonanTest.groovy:490]: Is this correct that we don't support modules in the two-stage mode?

File: build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt [hunk 2]
```diff
@@ -5,60 +5,114 @@
 
 package org.jetbrains.kotlin
 
+import java.nio.file.Path
 import java.nio.file.Paths
 import java.util.regex.Pattern
 
+private const val MODULE_DELIMITER = ",\\s*"
+private val FILE_OR_MODULE_PATTERN: Pattern = Pattern.compile("(?://\\s*MODULE:\\s*([^()\\n]+)(?:\\(([^()]+(?:" +
```
[the remark this answers is not in the document]
Comment (PavelPunegov) [reply build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt:13]: Well, I've got this pattern from the Kotlin repo for unification. And I don't know if it's intentional or not.

File: build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt [hunk 3]
```diff
@@ -5,60 +5,114 @@
...
+            val filePath = "$outputDirectory/$fileName"
             val start = processedChars
-            val nextFileExists = matcher.find()
+            nextFileExists = matcher.find()
             val end = if (nextFileExists) matcher.start() else srcText.length
             val fileText = srcText.substring(start, end)
             processedChars = end
-            registerKtFile(result, filePath, fileText)
-            if (!nextFileExists) break
+            if (fileName.endsWith(".kt")) {
```
[the remark this answers is not in the document]
Comment (PavelPunegov) [reply build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt:63]: There are certain tests that have .java files and some .smap files

File: build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt [hunk 4]
```diff
@@ -5,60 +5,114 @@
...
+            val moduleDependencies = matcher.group(2)
+            val moduleFriends = matcher.group(3)
+
+            if (moduleName != null) {
+                moduleName = moduleName.trim { it <= ' ' }
+                module = TestModule("${srcFile.name}.$moduleName",
+                        moduleDependencies.parseModuleList().map {
+                            if (it != "support") "${srcFile.name}.$it" else it
+                        },
+                        moduleFriends.parseModuleList().map { "${srcFile.name}.$it" })
```
[the remark this answers is not in the document]
Comment (PavelPunegov) [reply build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt:53]: What does being a friend module give to the main module? I'll try to use it as a friend.

File: build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilmat192) [reply build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt:13]: I see, thank you. May be add a comment with a link to this pattern in the Big Kotlin repo?

File: build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ilmat192) [reply build-tools/src/main/kotlin/org/jetbrains/kotlin/TestDirectives.kt:53]: > What does being a friend module give to the main module?

Oh, I see, the `support` module is compiled together with the main one ([here](https://github.com/JetBrains/kotlin-native/pull/4070/files#diff-212919446af9dbd167848d12a92d3812R517))? In this case the question is resolved.

File: build-tools/src/main/groovy/org/jetbrains/kotlin/KonanTest.groovy (hunk 1 above)
Comment (ilmat192) [reply build-tools/src/main/groovy/org/jetbrains/kotlin/KonanTest.groovy:491]: Discussed offline, decided to do this in a separate PR.

**Pull request merged**